src/hooks/useUserStats.ts

```diff
@@ -0,0 +1,131 @@
+      globalSharesSec
+      createdTimestamp
+    }
+    users(where: { id: $userAddress }) {
+      id
+      earned
+      stakes {
+        amount
+        timestamp
+      }
+    }
+  }
+`;
+
+type BigNumberish = BigNumber.Value;
+
+type Stake = {
+  amount: BigNumber;
+  shares: BigNumber;
+  timestamp: number;
+};
+
+type UserStatsResponse = {
+  earned: BigNumberish;
+  stakes: [Stake];
+};
+
+type UserStatsReturnType = {
+  earned: BigNumber;
+  stakes: [Stake];
+  totalStaked: BigNumber;
+  ownershipShare: number;
+  weightedMultiplier: number;
+};
+
+export function useUserStats(userAddress: string, farmAddress: string): UserStatsReturnType {
+  const [earned, setEarned] = React.useState(ZERO);
+  const [stakes, setStakes] = React.useState<[Stake]>([{ amount: ZERO, shares: ZERO, timestamp: 0 }]);
+  const [totalStaked, setTotalStaked] = React.useState(ZERO);
+  const [ownershipShare, setOwnershipShare] = React.useState(0);
+  const [weightedMultiplier, setWeightedMultiplier] = React.useState(0);
+
+  React.useEffect(() => {
+    let cancelled = false;
+    let retryTimer: ReturnType<typeof setTimeout>;
+
+    async function fetchUserStats() {
+      try {
+        const result = await graphqlClient.query(USER_STATS_QUERY, { userAddress, farmAddress }).toPromise();
+        if (!result?.data) {
+          return;
+        }
+
+        const [{ earned, stakes }]: [UserStatsResponse] = result.data.users;
+
+        const [{ globalSharesSec, bonusPeriodSec, createdTimestamp }]: [
+          { globalSharesSec: BigNumber; bonusPeriodSec: number; createdTimestamp: number },
+        ] = result.data.tokenGeysers;
+
+        const today = dayjs();
+        const parsedEarned = new BigNumber(earned);
+
+        let totalStakeShareSecs = ZERO;
+        let parsedTotalStaked = ZERO;
+        for (let index = 0; index < stakes.length; index++) {
+          const stake: Stake = stakes[index];
+          stake.amount = new BigNumber(stake.amount);
+          parsedTotalStaked = parsedTotalStaked.plus(stake.amount);
+          totalStakeShareSecs = totalStakeShareSecs.plus(stake.amount.times(today.unix() - stake.timestamp));
+        }
+
+        const calculatedOwnershipShare = totalStakeShareSecs.div(globalSharesSec).times(new BigNumber(100));
+        const parsedStakes = stakes;
+
+        const maxBonusDateSeconds = +createdTimestamp + +bonusPeriodSec;
+        const maxBonusDate = dayjs.unix(maxBonusDateSeconds);
+
+        let averageMultiplier = 0;
+        for (let index = 0; index < stakes.length; index++) {
+          const stake: Stake = stakes[index];
+          const stakeWeight = stake.amount.div(parsedTotalStaked).toNumber();
+
+          const stakeDate = dayjs.unix(stake.timestamp);
+          const secondsUntilMaxBonus = maxBonusDate.diff(stakeDate, 'seconds');
+          const weightedBonusMultiplier = ((secondsUntilMaxBonus * MAX_MULTIPLIER) / bonusPeriodSec) * stakeWeight;
+          averageMultiplier += weightedBonusMultiplier;
```

Comment:
If we're doing this loop for just calculating the whole multiplied, we could probably just do `.reduce()` on the array. 😄